diff --git a/src/utils/input_bit_stream.cc b/src/utils/input_bit_stream.cc
index 407487c..307e2f4 100644
--- a/src/utils/input_bit_stream.cc
+++ b/src/utils/input_bit_stream.cc
@@ -1,7 +1,7 @@
 #include "utils/input_bit_stream.h"
 
 InputBitStream::InputBitStream(uint8_t *raw_data, size_t size) {
-  data_ = Array<uint32_t>(std::ceil(static_cast<double>(size) / sizeof(uint32_t)));
+  data_ = Array<uint32_t>(std::ceil(static_cast<double>(size) / sizeof(uint32_t)) + 1);
   __builtin_memcpy(data_.begin(), raw_data, size);
   for (auto &blk : data_) blk = be32toh(blk);
   buffer_ = (static_cast<uint64_t>(data_[0])) << 32;
@@ -14,11 +14,17 @@ uint64_t InputBitStream::Peek(size_t len) {
 }
 
 void InputBitStream::Forward(size_t len) {
-  bit_in_buffer_ -= len;
-  buffer_ <<= len;
-  uint64_t load_more_bits = (bit_in_buffer_ < 32) * ((cursor_ < data_.length()) ? 1 : 0);
-  uint64_t next_bits = load_more_bits * data_[cursor_];
-  buffer_ |= next_bits << (32 - bit_in_buffer_);
-  bit_in_buffer_ += load_more_bits * 32;
-  cursor_ += load_more_bits;
+  if (len >= bit_in_buffer_) {
+    bit_in_buffer_ = 0;
+    buffer_ = 0;
+  } else {
+    bit_in_buffer_ -= len;
+    buffer_ <<= len;
+  }
+  if (bit_in_buffer_ < 32 && cursor_ < data_.length()) {
+    const uint64_t next_bits = data_[cursor_];
+    buffer_ |= next_bits << (32 - bit_in_buffer_);
+    bit_in_buffer_ += 32;
+    ++cursor_;
+  }
 }
@@ -45,7 +45,7 @@ uint32_t InputBitStream::ReadBit() {
 }
 
 void InputBitStream::SetBuffer(const Array<uint8_t> &new_buffer) {
-  data_ = Array<uint32_t>(std::ceil(static_cast<double>(new_buffer.length()) / sizeof(uint32_t)));
+  data_ = Array<uint32_t>(std::ceil(static_cast<double>(new_buffer.length()) / sizeof(uint32_t)) + 1);
   __builtin_memcpy(data_.begin(), new_buffer.begin(), new_buffer.length());
   for (auto &blk : data_) blk = be32toh(blk);
   buffer_ = (static_cast<uint64_t>(data_[0])) << 32;
@@ -54,7 +54,7 @@ void InputBitStream::SetBuffer(const Array<uint8_t> &new_buffer) {
 }
 
 void InputBitStream::SetBuffer(const std::vector<uint8_t> &new_buffer) {
-  data_ = Array<uint32_t>(std::ceil(static_cast<double>(new_buffer.size()) / sizeof(uint32_t)));
+  data_ = Array<uint32_t>(std::ceil(static_cast<double>(new_buffer.size()) / sizeof(uint32_t)) + 1);
   __builtin_memcpy(data_.begin(), new_buffer.data(), new_buffer.size());
   for (auto &blk : data_) blk = be32toh(blk);
   buffer_ = (static_cast<uint64_t>(data_[0])) << 32;
